contracts/yellow/VaultImpl.sol

```diff
@@ -1,11 +1,344 @@
+        require(recoveredSigner == _brokerVirtualAddress, 'Signer is not broker');
+
+        address newBrokerVirtualAddress = abi.decode(encodedAddress, (address));
+        require(newBrokerVirtualAddress != address(0),'Invalid virtual address');
+        _brokerVirtualAddress = newBrokerVirtualAddress;
+        emit BrokerVirtualAddressSet(newBrokerVirtualAddress);
+    }
+
+    /**
+     * @notice Set the address derived from the soSigner's new public key. Emits `CoSignerVirtualAddressSet` event.
+     * @dev Supplied payload must be signed by soSigner's current public key.
+     * @param encodedAddress Encoded new virtual soSigner address.
+     * @param signature New virtual address signed by soSigner's current public key.
+     */
+    function setCoSignerVirtualAddress(bytes memory encodedAddress, bytes calldata signature)
+        external
+    {
+        bytes32 digest = ECDSA.toEthSignedMessageHash(keccak256(encodedAddress));
+        address recoveredSigner = ECDSA.recover(digest, signature);
+        require(recoveredSigner == _coSignerVirtualAddress, 'Signer is not coSigner');
+
+        address newCoSignerVirtualAddress = abi.decode(encodedAddress, (address));
+        require(newCoSignerVirtualAddress != address(0),'Invalid virtual address');
+        _coSignerVirtualAddress = newCoSignerVirtualAddress;
+        emit CoSignerVirtualAddressSet(newCoSignerVirtualAddress);
+    }
+
+    /**
+     * @notice Get broker virtual (only public key it is derived from exists) key for this vault.
+     * @dev Get broker virtual (only public key it is derived from exists) key for this vault.
+     * @return address Broker virtual (only public key it is derived from exists) key.
+     */
+    function getBrokerVirtualAddress() external view returns (address) {
+        return _brokerVirtualAddress;
+    }
+
+    /**
+     * @notice Get coSigner virtual (only public key it is derived from exists) key for this vault.
+     * @dev Get coSigner virtual (only public key it is derived from exists) key for this vault.
+     * @return address CoSigner virtual (only public key it is derived from exists) key.
+     */
+    function getCoSignerVirtualAddress() external view returns (address) {
+        return _coSignerVirtualAddress;
+    }
+
+    /**
+     * @notice Deposit assets with given payload from the caller. Emits `Deposited` event.
+     * @dev Deposit assets with given payload from the caller. Emits `Deposited` event.
+     * @param encodedPayload Encoded payload, which denotes action to be performed.
+     * @param brokerSignature Payload signed by the broker.
+     * @param coSignerSignature Payload signed by the coSigner.
+     */
+    function deposit(
+        bytes calldata encodedPayload,
+        bytes calldata brokerSignature,
+        bytes calldata coSignerSignature
+    ) external payable {
+        address issuer = msg.sender;
+
+        // check signatures
+        _requireSigNotUsed(issuer, brokerSignature);
```

Comment:
do we also check that the rid was not processed yet?

Reply:
This is not needed, as these tests shows: https://github.com/yellorg/smart/pull/7/commits/b50ed7bfb3feacd83d32e8728d76cda946d8a293

This happen since signature check already act as a "already used" payload data